[IAC-927] Blog Post - Testing Puppet Modules with Litmus and Vagrant (#95)
puppetlabs/iac: _docs/LitmusVagrantTesting.md

---
layout: page
title: Testing Puppet Modules with Litmus and Vagrant
---

## Testing Puppet Modules with Litmus and Vagrant
 
### The Setup 

> For this demo I'm gonna run through the entire process from start to finish, I hope that you find it useful.

Reviewer: ```suggestion
For this demo I'm gonna run through the entire process from start to finish - I hope that you find it useful.
```

Author: Done

> To start with the first thing that you need to do is clone down the module that you want to test, in this case we'll be using `puppetlabs-scheduled_task`. This can easily be accomplished with a simple command:

Reviewer: ```suggestion
To start, the first thing that you need to do is clone down the module that you want to test; in this case we'll be using `puppetlabs-scheduled_task`. This can easily be accomplished with a simple command:
```

Author: done

```
git clone https://github.com/user-name/puppetlabs-scheduled_task.git
```

> Now that you have the module cloned down simply run the below command to set up the environment:

Reviewer: ```suggestion
Now that you have the module cloned down, simply run the below command to set up the environment:
```

Author: Done

```
bundle install --path .bundle/gems/
```

### Provisioning the Machine

> Now that the environment is setup it is time to provision the machine, while normally you would do this with a vagrant command, Litmus can do it for you if provided with the correct information, as shown below:

Reviewer 2: Can you replace the word "now" ? It's repeated here

Author: Changed to 'With the environment'

> Now that the environment is setup it is time to provision the machine, while normally you would do this with a vagrant command, Litmus can do it for you if provided with the correct information, as shown below:

Reviewer: ```suggestion
Now that the environment is setup, it's time to provision the machine.
Normally you would do this with a Vagrant command, but Litmus can do it for you, if provided with the correct information, as shown below:
```

Author: Changed to `With the environment is setup it is time to provision the machine. Rather than spin up the vagrant machine directly, we can use Litmus as an intermediary to do it for us.z

Reviewer: Just noticed a typo in this.

![Litmus Vagrant Testing - provisioning the machine](Vagrant%20Provision.png)

> If you are testing against multiple different versions then you can also utilise the provision_list command, which let's you set out an entire suite of OS to test against by adding them to and then calling them from the `provision.yml`.

Reviewer: ```suggestion
If you are testing against multiple different versions then you can also utilise the `provision_list` command, which adds/appends an entire suite of OSs to the `provision.yml`.
```

Author: Changed to: 
If you are testing against multiple different OSs then you can also utilise the `provision_list` command, which alows you to provision an entire suite of machines from pre-set lists found within the `provision.yml` file.

> With the machine provisioned it is a simple matter of running the `install_agent` and `install_module` commands and you will be setup and ready to go. With `install_agent` you can run the command as it is to install a default version of the agent or you can specify puppet 5 or 6 specificilly.

Reviewer: ```suggestion
With the machine provisioned it is simply a matter of running the `install_agent` and `install_module` commands and you will be setup and ready to go. With `install_agent` you can run the command as-is to install the default version of the agent, or you can specify `puppet5` or `puppet6` specifically.
```

![Litmus Vagrant Testing - installing the agent and the module](Agent%20Module%20Install.png)

### Running the Tests

Now that the machine is fully setup it is time to run the tests.

> This can be done in several way's, either through Litmus or through rspec, we'll start with the command `bundle exec rake litmus:acceptance:parallel`. 

Reviewer: ```suggestion
This can be done in several ways, either through Litmus or through rspec.
We'll start with the command `bundle exec rake litmus:acceptance:parallel`. 
```

Author: done

> When this command is given Litmus will retrieve each machine that has been setup and run the test suite against them simultaniusly. This is an excellent function for testing prior to a release or merge as it helps achieve full coverage in a fraction of the time that it would normally take.

Reviewer: ```suggestion
When this command is given, Litmus will retrieve each machine that has been setup during the provision step and run the test suite against them simultaneously. This is an excellent function for testing prior to a release or merge as it helps achieve full coverage in a fraction of the time that it would normally take.
```

Author: Done

> If however you are still in the development phase of your work or working on a fix for a specific OS then you may wish to test you changes against a single box. In which case you can do sby providing the name of the machine that you want to run your test's against, as shown here `rake litmus:acceptance:127.0.0.1:55985`.

Reviewer: ```suggestion
If, however you are still in the development phase of your work or working on a fix for a specific OS, then you may wish to test you changes against a single host, in which case you can do by providing the name of the machine that you want to run your tests against, as shown here `rake litmus:acceptance:127.0.0.1:55985`.
```

Author: Changed to:
The next command that can be given is `bundle exec rake litmus:acceptance:127.0.0.1:55985`

This command will then run the test's only against the machine whose name was provided, allowing you to run your test suite against a specific OS. This can be useful when making a change specific to a certain OS or when re-running previously failed test's after applying a quick bugfix.

> Finally if you wish to debug your test's with a little more detail or run a specific test file instead of the fully suite, you can use the older, but still functional, rspec command shown here `TARGET_HOST=127.0.0.1:55985 bundle exec rspec spec/acceptance/should_create_task_spec.rb`.

Reviewer: ```suggestion
Finally if you wish to debug your tests with a little more detail or run a specific test file instead of the full suite, you can use the older, but still functional, rspec command shown here `TARGET_HOST=127.0.0.1:55985 bundle exec rspec spec/acceptance/should_create_task_spec.rb`.
```

Author: Changed to:
The final command that we wll be covering is `TARGET_HOST=127.0.0.1:55985 bundle exec rspec spec/acceptance/should_create_task_spec.rb`.

Rather than calling on Litmus this instead utilise the older, but still functional, rspec command. This can allow you to specify specific test files to be run rather than the entire suite as well as allowing more detailed debugs to be returned.

### Wrapup

> Well, that's it, that's everything you need to know to get started with a vagrant test suite. I hope that you all find it helpful going forward and can all learn to experiance the joys of Vagrant.

Reviewer: ```suggestion
That's everything you need to know to get started with a vagrant test suite! I hope that you all find it helpful going forward and can all learn to experience the joys of Vagrant.
```

Author: Changed to:
Well that's it, everything you need to know to get started with a vagrant test suite! I hope that you all find it helpful going forward and can all learn to experience the joys of Vagrant.

---
layout: page
title: Testing Puppet Modules with Litmus and Vagrant
---

## Testing Puppet Modules with Litmus and Vagrant
 
### The Setup 

For this demo I'm gonna run through the entire process from start to finish - I hope that you find it useful.
To start, the first thing that you need to do is clone down the module that you want to test; in this case we'll be using `puppetlabs-scheduled_task`. This can easily be accomplished with a simple command:

```
git clone https://github.com/user-name/puppetlabs-scheduled_task.git
```

Now that you have the module cloned down, simply run the below command to set up the environment:

```
bundle install --path .bundle/gems/
```

### Provisioning the Machine

> With the environment is setup it is time to provision the machine. Rather than spin up the vagrant machine directly, we can use Litmus as an intermediary to do it for us.

Reviewer: ```suggestion
With the environment setup it is time to provision the machine. Rather than spin up the vagrant machine directly, we can use Litmus as an intermediary to do it for us.
```

Author: Done

With the machine provisioned, it's a simple matter of running the `install_agent` and `install_module` commands and you will be setup and ready to go. With `install_agent` you can run the command as it is to install a default version of the agent or you can specify puppet 5 or 6 specificilly.

![Litmus Vagrant Testing - installing the agent and the module](Agent%20Module%20Install.png)

### Running the Tests

Now that the machine is fully setup it is time to run the tests.

This can be done in several ways, either through Litmus or through rspec.

We'll start with the command `bundle exec rake litmus:acceptance:parallel`.

When this command is given, Litmus will retrieve each machine that has been setup during the provision step and run the test suite against them simultaneously. This is an excellent function for testing prior to a release or merge as it helps achieve full coverage in a fraction of the time that it would normally take.

The next command that can be given is `bundle exec rake litmus:acceptance:127.0.0.1:55985`

> This command will run the test's only against the machine whose name was provided, allowing you to run your test suite against a specific OS. This can be useful when making a change specific to a certain OS or when re-running previously failed test's after applying a quick bugfix.

Reviewer: ```suggestion
This command will run the tests only against the machine whose name was provided, allowing you to run your test suite against a specific OS. This can be useful when making a change specific to a certain OS or when re-running previously failed tests after applying a quick bug fix.
```

Author: Done

The final command that we wll be covering is `TARGET_HOST=127.0.0.1:55985 bundle exec rspec spec/acceptance/should_create_task_spec.rb`.

> Rather than calling on Litmus this instead utilise the older, but still functional, rspec command. This can allow you to specify specific test files to be run rather than the entire suite as well as allowing more detailed debugs to be returned.

Reviewer: ```suggestion
Rather than calling on Litmus this instead utilises the older, but still functional, rspec command. This will allow you to specify specific test files to be run rather than the entire suite as well as allowing more detailed debugs to be returned.
```

Author: Changed to:
This allows

We'll start with the command `bundle exec rake litmus:acceptance:parallel`.

When this command is given, Litmus will retrieve each machine that has been setup during the provision step and run the test suite against them simultaneously. This is an excellent function for testing prior to a release or merge as it helps achieve full coverage in a fraction of the time that it would normally take.

The next command that can be given is `bundle exec rake litmus:acceptance:127.0.0.1:55985`

This command will run the tests only against the machine whose name was provided, allowing you to run your test suite against a specific OS. This can be useful when making a change specific to a certain OS or when re-running previously failed tests after applying a quick bug fix.

The final command that we wll be covering is `TARGET_HOST=127.0.0.1:55985 bundle exec rspec spec/acceptance/should_create_task_spec.rb`.

Rather than calling on Litmus this instead utilises the older, but still functional, rspec command. This allows you to specify specific test files to be run rather than the entire suite as well as allowing more detailed debugs to be returned.

### Troubleshooting

> A certain Vagrant boxes do not allow ssh root logins the vagrant user is used to login and sudo is used to execute privileged commands. By As the Puppet agent installation doesn't touch the systems' sudo secure_path configuration this can lead to errors when attepting to execute puppet cmmands on the system. If you encounter these errors a quick fix is to use the bolt task `provision::fix_secure_path`. This will update the secure_path configuration and solve any related errors you may be seeing.

Reviewer 3: ```suggestion
Certain Vagrant boxes do not allow ssh root logins, that is why the _vagrant_ user as part of the sudoers group is used to execute privileged commands. As the Puppet agent installation requires sudo privileges, there are some cases where the secure_path configuration for the puppet binary path is not set correctly, leading to errors when attempting to execute puppet commands on the system. If you encounter these errors, a quick fix is to use the `provision::fix_secure_path` Bolt Task. This will update the secure_path configuration and solve any related errors you may be seeing.
```

Author: Done

We'll start with the command `bundle exec rake litmus:acceptance:parallel`.

When this command is given, Litmus will retrieve each machine that has been setup during the provision step and run the test suite against them simultaneously. This is an excellent function for testing prior to a release or merge as it helps achieve full coverage in a fraction of the time that it would normally take.

The next command that can be given is `bundle exec rake litmus:acceptance:127.0.0.1:55985`

This command will run the tests only against the machine whose name was provided, allowing you to run your test suite against a specific OS. This can be useful when making a change specific to a certain OS or when re-running previously failed tests after applying a quick bug fix.

The final command that we wll be covering is `TARGET_HOST=127.0.0.1:55985 bundle exec rspec spec/acceptance/should_create_task_spec.rb`.

Rather than calling on Litmus this instead utilises the older, but still functional, rspec command. This allows you to specify specific test files to be run rather than the entire suite as well as allowing more detailed debugs to be returned.

### Additional Parameters

> Aside from the basic usage shown above there are a few additional parameters than can be passed through during the proision command, allowing some extended functionality.

Reviewer 3: ```suggestion
Aside from the basic usage shown above there are a few additional parameters than can be passed through as environment variables during the provision command, allowing some extended functionality.
```

Author: Updated
